File: learn-greek-easy-frontend/src/types/situation.ts
// --- Enums as union types ---

export type SituationStatus = 'draft' | 'partial_ready' | 'ready';
export type SituationDialogStatus = 'draft' | 'audio_ready' | 'exercises_ready';
export type SituationDescriptionStatus = 'draft' | 'audio_ready';
export type SituationDescriptionSourceType = 'original' | 'news';
export type SituationPictureStatus = 'draft' | 'generated';

// --- Nested child types (for detail response) ---

export interface DialogSpeaker {
  id: string;
  speaker_index: number;
  character_name: string;
  voice_id: string;
}

export interface WordTimestamp {
  word: string;
  start_ms: number;
  end_ms: number;
}

export interface DialogLine {
  id: string;
  line_index: number;
  speaker_id: string;
  text: string;
  start_time_ms: number | null;
  end_time_ms: number | null;
  word_timestamps: WordTimestamp[] | null;
}

export interface DialogNested {
  id: string;
  status: SituationDialogStatus;
  num_speakers: number;
  audio_duration_seconds: number | null;
  audio_url: string | null;
  created_at: string;
  speakers: DialogSpeaker[];
  lines: DialogLine[];
}

export interface DescriptionNested {
  id: string;
  text_el: string;
  text_el_a2: string | null;
  source_type: SituationDescriptionSourceType;
  status: SituationDescriptionStatus;
  audio_duration_seconds: number | null;
  audio_a2_duration_seconds: number | null;
  audio_url: string | null;
  audio_a2_url: string | null;
  word_timestamps: WordTimestamp[] | null;
  word_timestamps_a2: WordTimestamp[] | null;
  created_at: string;
}

export interface PictureNested {
  id: string;
  image_prompt: string;
  status: SituationPictureStatus;
  created_at: string;
}

// --- List item (matches SituationListItem Pydantic schema) ---

export interface SituationListItem {
  id: string;
  scenario_el: string;
  scenario_en: string;
  scenario_ru: string;
  status: SituationStatus;
  created_at: string;
  has_dialog: boolean;
  has_description: boolean;
  has_picture: boolean;
  has_dialog_audio: boolean;
  has_description_audio: boolean;
  description_timestamps_count: number;
}

// --- Response (matches SituationResponse) ---

export interface SituationResponse {
  id: string;
  scenario_el: string;
  scenario_en: string;
  scenario_ru: string;
  status: SituationStatus;
  created_at: string;
  updated_at: string;
}

// --- Detail response (extends SituationResponse with nested children) ---

export interface SituationDetailResponse extends SituationResponse {
  dialog: DialogNested | null;
  description: DescriptionNested | null;
  picture: PictureNested | null;
}

// --- Create payload (matches SituationCreate) ---

export interface SituationCreatePayload {
  scenario_el: string;
  scenario_en: string;
  scenario_ru: string;
}

// --- Update payload ---

export interface SituationUpdatePayload {
  scenario_el?: string;
  scenario_en?: string;
  scenario_ru?: string;
}

// --- List response (matches SituationListResponse) ---

export interface SituationListResponse {
  items: SituationListItem[];
  total: number;
  page: number;
  page_size: number;
  status_counts: Record<string, number>;
}

// --- Learner-facing types (matches learner_situation.py schemas) ---

export interface LearnerDescriptionNested {
  text_el: string;
  text_el_a2: string | null;
  audio_url: string | null;
  audio_a2_url: string | null;
  audio_duration_seconds: number | null;
  audio_a2_duration_seconds: number | null;
  word_timestamps: WordTimestamp[] | null;
  word_timestamps_a2: WordTimestamp[] | null;
}

export interface LearnerDialogNested {
  speakers: DialogSpeaker[];
  lines: DialogLine[];
  audio_url: string | null;
  audio_duration_seconds: number | null;
}

export interface LearnerSituationListItem {
  id: string;
  scenario_el: string;
  scenario_en: string;
  scenario_ru: string;
  status: SituationStatus;
  has_audio: boolean;
  has_dialog: boolean;
  exercise_total: number;
  exercise_completed: number;
  source_image_url: string | null;
}

export interface LearnerSituationListResponse {
  items: LearnerSituationListItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface LearnerSituationDetailResponse {
  id: string;
  scenario_el: string;
  scenario_en: string;
  scenario_ru: string;
  status: SituationStatus;
  description: LearnerDescriptionNested | null;
  dialog: LearnerDialogNested | null;
  exercise_total: number;
  exercise_completed: number;
  source_url: string | null;
  source_image_url: string | null;
  source_title: string | null;
}

// Exercise types for admin situation detail
export type ExerciseType = 'fill_gaps' | 'select_heard' | 'true_false' | 'select_correct_answer';
export type ExerciseStatus = 'draft' | 'approved';
export type DeckLevel = 'A1' | 'A2' | 'B1' | 'B2';
export type ExerciseModality = 'listening' | 'reading';

export interface SituationExerciseItemResponse {
  item_index: number;
  payload: Record<string, unknown>;
}

export interface SituationExerciseResponse {
  id: string;
  exercise_type: ExerciseType;
  status: ExerciseStatus;
  items: SituationExerciseItemResponse[];
  audio_level?: DeckLevel;
  modality?: ExerciseModality;
  audio_url?: string;
  reading_text?: string;
}

export interface SituationExerciseGroupResponse {
  source_type: string;
  exercises: SituationExerciseResponse[];
  exercise_count: number;
}

export interface SituationExercisesResponse {
  groups: SituationExerciseGroupResponse[];
  total_count: number;
}

export type ExerciseSourceType = 'description' | 'dialog' | 'picture';

export interface AdminExerciseListItem {
  id: string;
  exercise_type: ExerciseType;
  status: ExerciseStatus;
  source_type: ExerciseSourceType;
  modality: ExerciseModality;
  audio_level: DeckLevel | null;
  situation_id: string;
  situation_title_el: string;
  situation_title_en: string;
  audio_url: string | null;
  reading_text: string | null;
  item_count: number;
  items: SituationExerciseItemResponse[];
}

export interface AdminExerciseListResponse {
  items: AdminExerciseListItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface AdminExerciseListParams {
  modality: 'listening' | 'reading';
  page?: number;
  page_size?: number;
  exercise_type?: string;
  status?: string;
  search?: string;
}

File: learn-greek-easy-backend/src/schemas/situation.py
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from src.db.models import (
    DeckLevel,
    DescriptionSourceType,
    DescriptionStatus,
    DialogStatus,
    ExerciseModality,
    ExerciseSourceType,
    ExerciseStatus,
    ExerciseType,
    PictureStatus,
    SituationStatus,
)
from src.schemas.admin import DialogLineDetail, DialogSpeakerDetail, WordTimestamp


class SituationCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    scenario_el: str = Field(min_length=1, max_length=500)
    scenario_en: str = Field(min_length=1, max_length=500)
    scenario_ru: str = Field(min_length=1, max_length=500)


class SituationUpdate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    scenario_el: Optional[str] = Field(default=None, min_length=1, max_length=500)
    scenario_en: Optional[str] = Field(default=None, min_length=1, max_length=500)
    scenario_ru: Optional[str] = Field(default=None, min_length=1, max_length=500)

    @model_validator(mode="after")
    def check_at_least_one_field(self) -> "SituationUpdate":
        if not self.model_fields_set:
            raise ValueError("At least one field must be provided for update")
        for field_name in self.model_fields_set:
            if getattr(self, field_name) is None:
                raise ValueError(f"{field_name} cannot be null")
        return self


class PictureUpdate(BaseModel):
    """Schema for updating a Picture's scene + style fields.

    Body-shape trio rule: the three scene_* fields must arrive together.
    Either all three are populated (after trim) to set scenes, or all three
    are explicitly empty/null to clear them. Mixed shapes are rejected so
    the FE form can rely on a single trio-paired control.

    `style_en` is independent and may be present or omitted on its own.
    """

    model_config = ConfigDict(str_strip_whitespace=True)

    scene_en: Optional[str] = Field(default=None, max_length=1000)
    scene_el: Optional[str] = Field(default=None, max_length=1000)
    scene_ru: Optional[str] = Field(default=None, max_length=1000)
    style_en: Optional[str] = Field(default=None, max_length=1000)

    @model_validator(mode="after")
    def check_at_least_one_field(self) -> "PictureUpdate":
        if not self.model_fields_set:
            raise ValueError("At least one field must be provided for update")
        return self

    @model_validator(mode="after")
    def validate_scene_trio_shape(self) -> "PictureUpdate":
        scene_keys = {"scene_en", "scene_el", "scene_ru"}
        present = scene_keys & self.model_fields_set
        # Partial presence (1 or 2 of 3) is always rejected.
        if 0 < len(present) < 3:
            raise ValueError("scene_en, scene_el and scene_ru must all be provided or all omitted")
        if len(present) == 3:
            # str_strip_whitespace already trimmed; treat None or "" as empty.
            populated = [bool(getattr(self, k)) for k in ("scene_en", "scene_el", "scene_ru")]
            # All three must be uniformly populated OR uniformly empty.
            if any(populated) and not all(populated):
                raise ValueError(
                    "scene_en, scene_el and scene_ru must all be provided or all omitted"
                )
        return self


class SituationListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    scenario_el: str
    scenario_en: str
    scenario_ru: str
    status: SituationStatus
    created_at: datetime
    has_dialog: bool
    has_description: bool
    has_picture: bool
    has_dialog_audio: bool
    has_description_audio: bool
    description_timestamps_count: int
    dialog_exercises_count: int = 0
    description_exercises_count: int = 0
    picture_exercises_count: int = 0


class SituationListResponse(BaseModel):
    items: list[SituationListItem]
    total: int
    page: int
    page_size: int
    status_counts: dict[str, int] = {}


class SituationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    scenario_el: str
    scenario_en: str
    scenario_ru: str
    status: SituationStatus
    created_at: datetime
    updated_at: datetime


class DialogNested(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    status: DialogStatus
    num_speakers: int
    audio_duration_seconds: float | None
    audio_url: str | None = None
    created_at: datetime
    speakers: list[DialogSpeakerDetail]
    lines: list[DialogLineDetail]


class DescriptionNested(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    text_el: str
    text_el_a2: str | None
    source_type: DescriptionSourceType
    status: DescriptionStatus
    audio_duration_seconds: float | None
    audio_a2_duration_seconds: float | None
    audio_url: str | None = None
    audio_a2_url: str | None = None
    word_timestamps: list[WordTimestamp] | None = None
    word_timestamps_a2: list[WordTimestamp] | None = None
    created_at: datetime


class PictureNested(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    image_prompt: str
    status: PictureStatus
    created_at: datetime
    scene_en: str | None = None
    scene_el: str | None = None
    scene_ru: str | None = None
    style_en: str | None = None


class SituationDetailResponse(SituationResponse):
    dialog: DialogNested | None = None
    description: DescriptionNested | None = None
    picture: PictureNested | None = None


class SituationExerciseItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    item_index: int = Field(..., description="Zero-based position of this item")
    payload: dict = Field(..., description="Exercise-type-specific JSONB content")


class SituationExerciseResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    exercise_type: ExerciseType
    status: ExerciseStatus
    items: list[SituationExerciseItemResponse]
    audio_level: DeckLevel | None = None
    modality: ExerciseModality | None = None
    audio_url: str | None = None
    reading_text: str | None = None


class SituationExerciseGroupResponse(BaseModel):
    source_type: ExerciseSourceType
    exercises: list[SituationExerciseResponse]
    exercise_count: int = Field(..., ge=0, description="Number of exercises in this group")


class SituationExercisesResponse(BaseModel):
    groups: list[SituationExerciseGroupResponse]
    total_count: int = Field(..., ge=0, description="Total exercises across all groups")


class AdminExerciseListItem(BaseModel):
    """A single exercise in the flat admin exercise list."""

    id: UUID
    exercise_type: ExerciseType
    status: ExerciseStatus
    source_type: ExerciseSourceType
    modality: ExerciseModality
    audio_level: DeckLevel | None = None
    situation_id: UUID
    situation_title_el: str
    situation_title_en: str
    audio_url: str | None = None
    reading_text: str | None = None
    item_count: int
    items: list[SituationExerciseItemResponse]


class AdminExerciseListResponse(BaseModel):
    items: list[AdminExerciseListItem]
    total: int
    page: int
    page_size: int
